### ISSUE TITLE
Input is cached when not immediately consumed by a State

### ISSUE THREAD

**Issue by ciro-unity** [UnityTechnologies/open-project-1#297]
**Short description**
Input is cached when not immediately consumed by a State in the state machine, and then used when the next state "interested" in that input is reached.

**Expected behaviour**
The input should simply be ignored, if not related to the state the machine is currently in.

**To Reproduce**
1. Open a gameplay scene, press Play.
2. Jump and while in the air, press the button to talk (called "ExtraAction" as of now).
3. Observe the character doesn't talk in the air (as expected).
4. Observe that, once landed, the character talks, even if the ExtraAction button is no longer pressed.

**Notes**
- The functionality described above to talk is a temporary one, but the issue might happen with any action that doesn't interrupt the walk.

**Comment by ciro-unity** [comment]
This also happens when attacking, and pressing the Attack input while the first attack is still happening.

**Comment by treivize** [comment]
Finally, I have the feeling that there are not lots of input which will create a problematic behavior, so what about consuming quickly the input in Protagonist script after activation like this?

![image](https://user-images.githubusercontent.com/42570903/104169394-cada4480-53ff-11eb-9f41-96a839896f19.png)

It is a bit hacky but it seems good enough for the problem we want to solve the problem here.

**Comment by ciro-unity** [comment]
Hmm, it does feel kinda hacky, but it's also kind of safe because all it does is reset a bool. If we go that route, I'd unify it and make it a constant on all actions that need to be instantaneous (I guess all of them).

**Comment by DeivSky** [comment]
Someone mentioned in the forums about doing it in `LateUpdate`, I think that would be the simplest solution.
``` cs
private void LateUpdate()
{
    attackInput = false;
    talkInput = false;
    otherInput = false;
}
```
Another option I just came up with, and I think the least hacky one -because it's listening to what we actually care about in the state we're in, instead of randomly setting things to false,- would be to make an **Action** to dynamically subscribe and unsubscribe from events.

Here is the way I imagine it: 

- We modify `InputReader.cs` to invoke **EventChannelSOs** instead of UnityActions.  In turn, we'd also have to modify `Protagonist.cs` to take in and subscribe to **EvenChannelSOs**. 
- We write the new `BindEventChannelsActionSO.cs`. An **Action** that takes in two **EventChannelSOs**. In `OnStateEnter()` we subscribe raising one of the channels to the other, and in `OnStateExit()` we unsubscribe it. See where this is going? 
- We then make two sets of **EventChannelSOs**, a set for `Protagonist.cs`, which would be the set that gets subscribed to the channels from the other set, the set for `InputReader.cs`.
- Then we just have to make the `ActionSO` assets, bind the `EventChannelSO`s, and add them to each `State`.
- We only have to do this for the inputs that require it, the rest can be left as is or, preferably, using the same channel in both scripts.

I know it's probably quite a bit of work for something that can be solved with the few lines above but, what do you think?

**Comment by ciro-unity** [comment]
> what do you think?

I will reply with the words of a wise man:

> it's probably quite a bit of work for something that can be solved with the few lines above

🤣

**Comment by DeivSky** [comment]
Hahaha, I could sense your head shaking as I was typing 🤣. I thought I'd share the idea anyway 😉.

**Comment by rainaldisatria** [comment]
Solved the issue by modifying Protagonist.cs to this:
![image](https://user-images.githubusercontent.com/57592497/105035320-5fdade80-5a8d-11eb-8e38-1006b8dbfa58.png)

So the the thing about consuming input by using IsHoldingAttackActionConditionSO is that...
![image](https://user-images.githubusercontent.com/57592497/105035548-bd6f2b00-5a8d-11eb-8eef-9a3bf8c0c3ee.png)

It'll only be called in the state that's checking that condition (IsHoldingAttackActionConditionSO). Only Idle has this condition. 

To be more clear:
(...) = State.
(Idle) -> Attack button is pressed -> consumed the input -> (Attack) -> immediately press the attack button -> ((should be consumed here, but it's not)) -> (Idle) -> attack button is still pressed because it's not consumed -> consumed the input -> (Attack)
*notice that after immediately press the attack button, the attack input is not consumed, because the Attack state doesn't check IsHoldingAttackActionConditionSO

So that is why the input is cached.

**Comment by rainaldisatria** [comment]
I'm thinking about another solutions, but seems like the best and clean way to consume the input is with that StartCoroutine way.

We could consume it OnStateExit of Attack, but it will only solve when we spamming the attack button. Not gonna help when you press Jump then pressing attack (Jump doesn't consume the attack)

So the only way to not use the StartCoroutine way is to consume Attack on every state by adding action ConsumeAttackOnStateExitAction to every state (Idle, JumpDescending, ...)

**Comment by DeivSky** [comment]
Hey @rainaldisatria, the thing is that this doesn't only happen when attacking, it happens when talking, picking things up, and any other 'tap' actions we might add in the future, so we'd need a coroutine for each. That's why I thought setting all those booleans to false in `LateUpdate` would be the simplest thing to do. 

Thinking more about it though, since these events come from the input system, I think that could lead us to a situation where it's set to true _after_ `Update` but then reset in `LateUpdate`, completely missing the Condition and thus not triggering the Transition. That would be an advantage of using coroutines, with `yield return null;` we make sure we go through `Update` before resetting the value. However, for simplicity, I'd still make only one coroutine where all values are reset, instead of one for each.

**Comment by AlexandreGheraibia** [comment]
Hi everyone,
My solution is to consume the inputs attackInput and extraActionInput in the function OnStateExit() of PlayLandParticlesActionSO

**Comment by AlexandreGheraibia** [comment]
For bypass the Attack problem, I enabled transition from a state to the same state in the class AddTransitionHelper.
And in Asset > ScriptableObjects > StateMachine > Protagonist > PigChef_TransitionTable, I Add a transition from Attack to Attack with the condition if IsHoldingAttack == False and Timer_AttackAnimation == True

For the Jump problem, I added  the condition if isHoldingAttack == true to the transition from JumpDescending to Idle.

**Comment by AlexandreGheraibia** [comment]
Hi,

In Asset > ScriptableObjects > StateMachine > Protagonist > PigChef_TransitionTable.asset.
I changed the Transition Attack to Idle by the following condition : 
if Timer_AttackAnimation is True
And if IsMoving is False
Or if TimerAttack is True 
And if IsHoldingAttack is True

For resolve the problem when you Jump, I changed the transition JumpDescending to Idle by the following condition : 

if IsGrounded is True
And if IsMoving is False
Or  if TimerJumpHoldButton is True
And if IsHoldingAttack is True

You can do the same change for the transitions to walking.

**Comment by maltrusmaltren** [comment]
Hi for all.

I made two little changes on State Machine.

1st change - Walking State:
I added a transition to Talk State when player will talk to someone and not in proccess of jump.
![image](https://user-images.githubusercontent.com/396147/106334771-18f99f80-626a-11eb-9237-e8acb5908c6c.png)


2nd change - Talking State:
I added more one action to disable player walking.
![image](https://user-images.githubusercontent.com/396147/106334790-24e56180-626a-11eb-9e3f-499d267c98fc.png)


I think that is expected behaviour that was reported by drbir in the forum.

In my tests this two changes has been put the train on the track.
:)

Please feel free to add or remove any change that I maded.

**Comment by AlexandreGheraibia** [comment]
my idea for prevent action from jumpDescending
![image](https://user-images.githubusercontent.com/27300700/106369643-787faa00-6353-11eb-83b9-cd4cff205dc9.png)

**Comment by ciro-unity** [comment]
@AlexandreGheraibia I provided some feedback and a question in #338

**Comment by DeivSky** [comment]
@AlexandreGheraibia I kind of get where you're going, but wow... you truly beat me at overcomplicating things there.
This issue has been open for longer than I expected, so I've made a simple fix (#339) using coroutines. I thought I could get away with using just one but I was actually unaware that the interaction input was in a separate script (haven't had much time to catch up with all the systems).
